packages/core/src/RenderPipeline/batcher/Batcher2D.ts

```diff
@@ -0,0 +1,175 @@
+    this._engine = engine;
+    this._maxVertexCount = maxVertexCount;
+  }
+
+  /**
+   * Destroy internal resources.
+   */
+  destroy(): void {
+    const { _meshBuffers } = this;
+    for (let i = 0, l = _meshBuffers.length; i < l; ++i) {
+      _meshBuffers[i].destroy();
+    }
+    _meshBuffers.length = 0;
+    this._meshBuffers = null;
+    this._engine = null;
+    this._preElement = null;
+  }
+
+  commitRenderElement(element: RenderElement): RenderElement | null {
+    const { _preElement: preElement } = this;
+    let batchElement = null;
+    if (preElement) {
+      if (this._canBatch(preElement, element)) {
+        this._udpateRenderData(preElement, element, true);
+      } else {
+        batchElement = this._preElement;
+        this._udpateRenderData(preElement, element, false);
+      }
+    } else {
+      this._udpateRenderData(preElement, element, false);
+    }
+
+    return batchElement;
+  }
+
+  uploadBuffer(): void {
+    const { _meshBuffers: meshBuffers } = this;
+    for (let i = 0, l = meshBuffers.length; i < l; ++i) {
+      meshBuffers[i].uploadBuffer();
+    }
+  }
+
+  clear() {
+    this._reset();
+    const { _meshBuffers: meshBuffers } = this;
+    for (let i = 0, l = meshBuffers.length; i < l; ++i) {
+      meshBuffers[i].clear();
+    }
+  }
+
+  allocateChunk(vertexCount: number): MBChunk | null {
+    const { _meshBuffers } = this;
+    const len = _meshBuffers.length;
+    let chunk: MBChunk = null;
+    for (let i = 0; i < len; ++i) {
+      chunk = _meshBuffers[i].allocateChunk(vertexCount);
+      if (chunk) {
+        chunk._mbId = i;
+        return chunk;
+      }
+    }
+
+    const meshBuffer = this._createMeshBuffer(len, this._maxVertexCount);
+    chunk = meshBuffer.allocateChunk(vertexCount);
+    if (chunk) {
+      chunk._mbId = len;
+      return chunk;
+    }
+    return null;
+  }
+
+  freeChunk(chunk: MBChunk): void {
+    const meshBuffer = this._meshBuffers[chunk._mbId];
+    meshBuffer && meshBuffer.freeChunk(chunk);
+  }
+
+  protected _createMeshBuffer(index: number, maxVertexCount: number = Batcher2D.MAX_VERTEX_COUNT): MeshBuffer {
+    return (this._meshBuffers[index] ||= new MeshBuffer(this._engine, maxVertexCount));
+  }
+
+  private _canBatch(preElement: RenderElement, cureElement: RenderElement): boolean {
+    const preRenderData = <SpriteRenderData>preElement.data;
+    const curRenderData = <SpriteRenderData>cureElement.data;
+    if (preRenderData.chunk._meshBuffer !== curRenderData.chunk._meshBuffer) {
+      return false;
+    }
+
+    const preRender = <SpriteRenderer>preRenderData.component;
+    const curRender = <SpriteRenderer>curRenderData.component;
+
+    // Compare mask.
+    if (!this._checkBatchWithMask(preRender, curRender)) {
+      return false;
+    }
+
+    // Compare texture.
+    if (preRenderData.texture !== curRenderData.texture) {
+      return false;
+    }
+
+    // Compare material.
+    return preRenderData.material === curRenderData.material;
```

Comment:
Still hasn’t solved the problem that `SpriteRenderer` with masks cannot be batched together.

Reply:
This PR doesn't change the previous logic